File: src/poster.py
import json
import logging
import os
import subprocess

logger = logging.getLogger(__name__)


class Poster:
    def __init__(self, twitter_env: dict[str, str]):
        self.env = {**os.environ, **twitter_env}

    def post_reply(self, tweet_id: str, text: str) -> bool:
        try:
            result = subprocess.run(["twitter", "reply", tweet_id, text], capture_output=True, text=True, env=self.env)
            if result.returncode != 0:
                logger.error("Failed to post reply: %s", result.stderr)
                return False
            data = json.loads(result.stdout)
            return data.get("ok", False)
        except Exception as e:
            logger.error("Error posting reply: %s", e)
            return False

    def check_auth(self) -> bool:
        try:
            result = subprocess.run(["twitter", "status", "--json"], capture_output=True, text=True, env=self.env)
            if result.returncode != 0:
                return False
            data = json.loads(result.stdout)
            return data.get("ok", False) and data.get("data", {}).get("authenticated", False)
        except Exception:
            return False

File: src/poster.ts
import { spawnSync } from 'child_process';
import { config } from './config';

export class Poster {
  private env: NodeJS.ProcessEnv;

  constructor() {
    this.env = {
      ...process.env,
      TWITTER_AUTH_TOKEN: config.twitter_auth_token,
      TWITTER_CT0: config.twitter_ct0,
    };
  }

  postReply(tweetId: string, text: string): boolean {
    try {
      const result = spawnSync('twitter', ['reply', tweetId, text], {
        env: this.env,
        encoding: 'utf-8',
        timeout: 30000,
      });

      if (result.status !== 0) {
        console.error('[ERROR] Failed to post reply:', result.stderr);
        return false;
      }

      return true;
    } catch (err) {
      console.error('[ERROR] Error posting reply:', err);
      return false;
    }
  }

  checkAuth(): boolean {
    try {
      const result = spawnSync('twitter', ['status', '--json'], {
        env: this.env,
        encoding: 'utf-8',
        timeout: 30000,
      });

      if (result.status !== 0) {
        return false;
      }

      const data = JSON.parse(result.stdout);
      return !!(data?.ok && data?.data?.authenticated);
    } catch {
      return false;
    }
  }
}

export const poster = new Poster();
